src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Core.cs

```diff
@@ -136,14 +138,35 @@ public DirectoryLoadContext(string directory, DefaultAnalyzerAssemblyLoader load
                     return _compilerLoadContext.LoadFromAssemblyName(assemblyName);
                 }
 
-                var pathToLoad = _loader.GetPathToLoad(assemblyPath);
-                return LoadFromAssemblyPath(pathToLoad);
+                Debug.Assert(paths.Any());
+                if (paths.Contains(assemblyPath))
+                {
+                    return LoadFromAssemblyPath(_loader.GetPathToLoad(assemblyPath));
+                }
+
+                // The assembly isn't expected to be found at 'assemblyPath',
+                // but some assembly with the same name is known to the loader.
+                foreach (var candidatePath in paths)
+                {
+                    // Note: we assume that the assembly really can be found at this path,
+                    // and that calling GetAssemblyName doesn't cause us to hold a lock on the file
+                    var candidateName = AssemblyName.GetAssemblyName(candidatePath);
+                    // checking FullName ensures that version and PublicKeyToken match exactly
+                    if (candidateName.FullName.Equals(assemblyName.FullName, StringComparison.OrdinalIgnoreCase))
```

Comment:
This could cause unexpected behavior. Consider for example if the analyzer had a reference to C at version 2. Now consider a scenario where other analyzers provide version 1 and version 3. It's a toss up to which would get loaded here. I think we should be considering the `AssemblyName` in the list here and then preferring an exact match first and lacking that use the highest available version.

Reply:
This FullName includes the version and therefore requires an exact match. I could replace it with explicit comparisons of the name, version and PKT if you prefer.

I think the described scenario is covered by one of the new tests.

Reply:
I misinterpreted the comment at first. Have now addressed the concern by searching for the candidate with the highest version.